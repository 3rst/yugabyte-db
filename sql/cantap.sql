\unset ECHO
\i test_setup.sql

SELECT plan(63);
CREATE SCHEMA someschema;
CREATE FUNCTION someschema.huh () RETURNS BOOL AS 'SELECT TRUE' LANGUAGE SQL;
--SELECT * FROM no_plan();

/****************************************************************************/
-- Test can_ok().
SELECT * FROM check_test(
    can_ok( 'now' ),
    true,
    'simple function',
    'Function now() should exist',
    ''
);

SELECT * FROM check_test(
    can_ok( 'pg_catalog', 'now'::name ),
    true,
    'simple schema.function',
    'Function pg_catalog.now() should exist',
    ''
);

SELECT * FROM check_test(
    can_ok( 'now', 'whatever' ),
    true,
    'simple function desc',
    'whatever',
    ''
);

SELECT * FROM check_test(
    can_ok( 'now', '{}'::name[] ),
    true,
    'simple with 0 args',
    'Function now() should exist',
    ''
);

SELECT * FROM check_test(
    can_ok( 'now', '{}'::name[], 'whatever' ),
    true,
    'simple with 0 args desc',
    'whatever',
    ''
);

SELECT * FROM check_test(
    can_ok( 'pg_catalog', 'now', '{}'::name[] ),
    true,
    'simple schema.func with 0 args',
    'Function pg_catalog.now() should exist',
    ''
);

SELECT * FROM check_test(
    can_ok( 'pg_catalog', 'now', 'whatever' ),
    true,
    'simple schema.func with desc',
    'whatever',
    ''
);

SELECT * FROM check_test(
    can_ok( 'pg_catalog', 'now', '{}'::name[], 'whatever' ),
    true,
    'simple scchma.func with 0 args, desc',
    'whatever',
    ''
);

SELECT * FROM check_test(
    can_ok( 'lower', '{text}'::name[] ),
    true,
    'simple function with 1 arg',
    'Function lower(text) should exist',
    ''
);

SELECT * FROM check_test(
    can_ok( 'decode', '{text,text}'::name[] ),
    true,
    'simple function with 2 args',
    'Function decode(text, text) should exist',
    ''
);

SELECT * FROM check_test(
    can_ok( 'array_cat', ARRAY['anyarray','anyarray'] ),
    true,
    'simple array function',
    'Function array_cat(anyarray, anyarray) should exist',
    ''
);

-- Check a custom function with an array argument.
CREATE FUNCTION __cat__ (text[]) RETURNS BOOLEAN
AS 'SELECT TRUE'
LANGUAGE SQL;

SELECT * FROM check_test(
    can_ok( '__cat__', '{text[]}'::name[] ),
    true,
    'custom array function',
    'Function __cat__(text[]) should exist',
    ''
);

-- Check a custom function with a numeric argument.
CREATE FUNCTION __cat__ (numeric(10,2)) RETURNS BOOLEAN
AS 'SELECT TRUE'
LANGUAGE SQL;

SELECT * FROM check_test(
    can_ok( '__cat__', '{numeric}'::name[] ),
    true,
    'custom numeric function',
    'Function __cat__(numeric) should exist',
    ''
);

-- Check failure output.
SELECT * FROM check_test(
    can_ok( '__cat__', '{varchar[]}'::name[] ),
    false,
    'failure output',
    'Function __cat__(varchar[]) should exist',
    '' -- No diagnostics.
);

/****************************************************************************/
-- Try can() function names.
SELECT * FROM check_test(
    can( 'pg_catalog', ARRAY['lower', 'upper'], 'whatever' ),
    true,
    'can(schema) with desc',
    'whatever',
    ''
);

SELECT * FROM check_test(
    can( 'pg_catalog', ARRAY['lower', 'upper'] ),
    true,
    'can(schema)',
    'Schema pg_catalog can',
    ''
);

SELECT * FROM check_test(
    can( 'pg_catalog', ARRAY['lower', 'foo', 'bar'], 'whatever' ),
    false,
    'fail can(schema) with desc',
    'whatever',
    '    pg_catalog.foo() missing
    pg_catalog.bar() missing'
);

SELECT * FROM check_test(
    can( 'someschema', ARRAY['huh', 'lower', 'upper'], 'whatever' ),
    false,
    'fail can(someschema) with desc',
    'whatever',
    '    someschema.lower() missing
    someschema.upper() missing'
);

SELECT * FROM check_test(
    can( ARRAY['__cat__', 'lower', 'upper'], 'whatever' ),
    true,
    'can() with desc',
    'whatever',
    ''
);

SELECT * FROM check_test(
    can( ARRAY['lower', 'upper'] ),
    true,
    'can(schema)',
    'Schema ' || array_to_string(current_schemas(true), ' or ') || ' can',
    ''
);

SELECT * FROM check_test(
    can( ARRAY['__cat__', 'foo', 'bar'], 'whatever' ),
    false,
    'fail can() with desc',
    'whatever',
    '    foo() missing
    bar() missing'
);


/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
