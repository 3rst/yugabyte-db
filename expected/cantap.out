\unset ECHO
1..105
ok 1 - simple function should pass
ok 2 - simple function should have the proper description
ok 3 - simple function should have the proper diagnostics
ok 4 - simple schema.function should pass
ok 5 - simple schema.function should have the proper description
ok 6 - simple schema.function should have the proper diagnostics
ok 7 - simple function desc should pass
ok 8 - simple function desc should have the proper description
ok 9 - simple function desc should have the proper diagnostics
ok 10 - simple with 0 args should pass
ok 11 - simple with 0 args should have the proper description
ok 12 - simple with 0 args should have the proper diagnostics
ok 13 - simple with 0 args desc should pass
ok 14 - simple with 0 args desc should have the proper description
ok 15 - simple with 0 args desc should have the proper diagnostics
ok 16 - simple schema.func with 0 args should pass
ok 17 - simple schema.func with 0 args should have the proper description
ok 18 - simple schema.func with 0 args should have the proper diagnostics
ok 19 - simple schema.func with desc should pass
ok 20 - simple schema.func with desc should have the proper description
ok 21 - simple schema.func with desc should have the proper diagnostics
ok 22 - simple scchma.func with 0 args, desc should pass
ok 23 - simple scchma.func with 0 args, desc should have the proper description
ok 24 - simple scchma.func with 0 args, desc should have the proper diagnostics
ok 25 - simple function with 1 arg should pass
ok 26 - simple function with 1 arg should have the proper description
ok 27 - simple function with 1 arg should have the proper diagnostics
ok 28 - simple function with 2 args should pass
ok 29 - simple function with 2 args should have the proper description
ok 30 - simple function with 2 args should have the proper diagnostics
ok 31 - simple array function should pass
ok 32 - simple array function should have the proper description
ok 33 - simple array function should have the proper diagnostics
ok 34 - custom array function should pass
ok 35 - custom array function should have the proper description
ok 36 - custom array function should have the proper diagnostics
ok 37 - custom numeric function should pass
ok 38 - custom numeric function should have the proper description
ok 39 - custom numeric function should have the proper diagnostics
ok 40 - failure output should fail
ok 41 - failure output should have the proper description
ok 42 - failure output should have the proper diagnostics
ok 43 - simple function should pass
ok 44 - simple function should have the proper description
ok 45 - simple function should have the proper diagnostics
ok 46 - simple schema.function should pass
ok 47 - simple schema.function should have the proper description
ok 48 - simple schema.function should have the proper diagnostics
ok 49 - simple function desc should pass
ok 50 - simple function desc should have the proper description
ok 51 - simple function desc should have the proper diagnostics
ok 52 - simple with 0 args should pass
ok 53 - simple with 0 args should have the proper description
ok 54 - simple with 0 args should have the proper diagnostics
ok 55 - simple with 0 args desc should pass
ok 56 - simple with 0 args desc should have the proper description
ok 57 - simple with 0 args desc should have the proper diagnostics
ok 58 - simple schema.func with 0 args should pass
ok 59 - simple schema.func with 0 args should have the proper description
ok 60 - simple schema.func with 0 args should have the proper diagnostics
ok 61 - simple schema.func with desc should pass
ok 62 - simple schema.func with desc should have the proper description
ok 63 - simple schema.func with desc should have the proper diagnostics
ok 64 - simple scchma.func with 0 args, desc should pass
ok 65 - simple scchma.func with 0 args, desc should have the proper description
ok 66 - simple scchma.func with 0 args, desc should have the proper diagnostics
ok 67 - simple function with 1 arg should pass
ok 68 - simple function with 1 arg should have the proper description
ok 69 - simple function with 1 arg should have the proper diagnostics
ok 70 - simple function with 2 args should pass
ok 71 - simple function with 2 args should have the proper description
ok 72 - simple function with 2 args should have the proper diagnostics
ok 73 - simple array function should pass
ok 74 - simple array function should have the proper description
ok 75 - simple array function should have the proper diagnostics
ok 76 - custom array function should pass
ok 77 - custom array function should have the proper description
ok 78 - custom array function should have the proper diagnostics
ok 79 - custom numeric function should pass
ok 80 - custom numeric function should have the proper description
ok 81 - custom numeric function should have the proper diagnostics
ok 82 - failure output should fail
ok 83 - failure output should have the proper description
ok 84 - failure output should have the proper diagnostics
ok 85 - can(schema) with desc should pass
ok 86 - can(schema) with desc should have the proper description
ok 87 - can(schema) with desc should have the proper diagnostics
ok 88 - can(schema) should pass
ok 89 - can(schema) should have the proper description
ok 90 - can(schema) should have the proper diagnostics
ok 91 - fail can(schema) with desc should fail
ok 92 - fail can(schema) with desc should have the proper description
ok 93 - fail can(schema) with desc should have the proper diagnostics
ok 94 - fail can(someschema) with desc should fail
ok 95 - fail can(someschema) with desc should have the proper description
ok 96 - fail can(someschema) with desc should have the proper diagnostics
ok 97 - can() with desc should pass
ok 98 - can() with desc should have the proper description
ok 99 - can() with desc should have the proper diagnostics
ok 100 - can(schema) should pass
ok 101 - can(schema) should have the proper description
ok 102 - can(schema) should have the proper diagnostics
ok 103 - fail can() with desc should fail
ok 104 - fail can() with desc should have the proper description
ok 105 - fail can() with desc should have the proper diagnostics
